File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Ruby/recover_binary_search_tree.rb
# Two elements of a binary search tree (BST) are swapped by mistake.
#
#   Recover the tree without changing its structure.
#
#   Note:
#   A solution using O(n) space is pretty straight forward. Could you devise a constant space solution?
#   confused what "{1,#,2,3}" means? > read more on how binary tree is serialized on OJ.


# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
class RecoverTree
  def initialize
    @first_node = nil
    @second_node = nil
    @prev_node = nil
  end

  def swap
    @first_node.val, @second_node.val = @second_node.val, @first_node.val
  end

  def inorder_traversal(node)
    if(!node)
      return
    end

    inorder_traversal(node.left)

    if !@prev_node
      @prev_node = node
    else
      if @prev_node.val > node.val
        if @first_node
          @second_node = node
        else
          @first_node = @prev_node
          @second_node = node
        end
      end
    end

    @prev_node = node

    inorder_traversal(node.right)
  end
end

def recover_tree(root)
  # inorder traversal get the two odd nodes
  inst = RecoverTree.new
  inst.inorder_traversal(root)
  inst.swap
  return
end


one = TreeNode.new(1)
two = TreeNode.new(2)
thr = TreeNode.new(3)

#one.right = two
#two.left = thr
one.left = two

def traverse(node)
  if !node
    return
  end

  traverse(node.left)
  p node.val
  traverse(node.right)
end

traverse(one)
recover_tree(one)
traverse(one)

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/RecoverBinarySearchTree.py
# definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# elements of a binary search tree (BST) are swapped by mistake.
# Recover the tree without changing its structure. 

class Solution:
    # @param root, a tree node
    # @return a tree node
    def recoverTree(self, root):
        """
        two nodes in binary search tree are swapped
        """

        # using inorder traversal to find odd node
        self.first_node, self.second_node = None, None
        self.prev = None
        self.inorder( root )

        # swap two nodes' value
        if self.second_node != None:
            print "found"
            temp_val = self.second_node.val
            self.second_node.val = self.first_node.val
            self.first_node.val = temp_val
        print "After swap"
        self.inorder( root)
        return root
    
    def inorder(self, node):
        if node == None:
            return
        
        self.inorder( node.left)
        
        if self.prev == None:
            self.prev = node
        elif self.prev.val > node.val:
            if self.first_node == None:
                self.first_node = self.prev
                self.second_node = node         # if only two nodes, and swapped
            else: 
                self.second_node = node         # encount second time
        else:
            self.prev = node 
        print node.val
        
        self.inorder( node.right)
            
if __name__ == "__main__":
    # test what if only two nodes in tree
    print "Test#1"
    node_1 = TreeNode(0)
    node_2 = TreeNode(1)
    node_1.left = node_2
    
    inst = Solution()
    inst.recoverTree( node_1 ) 

    print "Test#2"
    node_1.val = 2
    node_2.val = 1
    node_1.left = None
    node_1.right = node_2
    inst.recoverTree( node_1 ) 

    print "Test#3"
    node_1.val = 3
    node_2.val = 1
    node_3 = TreeNode(2)
    node_1.left = node_2
    node_1.right = node_3
    inst.recoverTree( node_1)
